Gutenblocks: Subscription form block

dereksmart opened the pull request:
Even more scoped version of our gutenblocks.  I was seeing a lot of errors with the shortcode blocks (recipe and quiz), so maybe we can add those in a separate PR.  

With Gutenberg active, you should see a block for Subscription form.  

To Test: 
- Activate Gutenberg, activate Jetpack shortcodes and Subscription form feature.  
- You should see the options to add blocks for the mentioned above. 
- Make sure things work well
- Make sure it looks good on the front end
- Make sure there are no errors when Gutenberg is not active or installed.

zinigor commented:
Just a minor suggestion. What do you think about aligning the label to the left like this?
![screenshot from 2017-10-20 14-09-14](https://user-images.githubusercontent.com/374293/31818373-817c8f6a-b5a0-11e7-9829-3affb8afdff7.png)

zinigor commented:
One more problem: tab order is broken - if you try to tab your way through the form, after the description you end up outside the block.

zinigor commented on gulpfile.babel.js:
```diff
@@ -493,14 +495,38 @@ gulp.task( 'languages:extract', function( done ) {
+ * Gutenpack!
+ */
+gulp.task( 'gutenpack', function() {
+	gulp.src( '**/*/*block.jsx' )
```
While this is really minor, and we don't do it everywhere, let's `return` the result here so that gulp tracks the task completion properly. Without it we get:
```
[13:59:36] Starting 'gutenpack'...
[13:59:36] Finished 'gutenpack' after 836 μs
```

And when returning the result we get:
```
[14:22:03] Starting 'gutenpack'...
[14:22:05] Finished 'gutenpack' after 2.43 s
```

zinigor commented on gulpfile.babel.js:
```diff
@@ -493,14 +495,38 @@ gulp.task( 'languages:extract', function( done ) {
+} );
+
+gulp.task( 'gutenpack:watch', function() {
+	gulp.watch( [ '**/*/*block.jsx' ], [ 'gutenpack' ] );
```
Same here as in the task above.

zinigor commented on modules/subscriptions.php:
```diff
@@ -741,11 +756,48 @@ function set_cookies( $subscribe_to_post = false, $post_id = null, $subscribe_to
+			'Subscribe' => __( 'Subscribe', 'jetpack' ),
+			"Success! An email was just sent to confirm your subscription. Please find the email now and click 'Confirm Follow' to start subscribing."
+				=> __( "Success! An email was just sent to confirm your subscription. Please find the email now and click 'Confirm Follow' to start subscribing.", 'jetpack' ),
+			'Join %s other subscribers' => __( 'Join %s other subscribers', 'jetpack' ),
```
Sorry, but here we really need to use `_n`. I'd say we just return a string from the subscriber count endpoint, rather than a number.

dereksmart commented:
@zinigor thanks for the review, I've addressed your feedback!

> Just a minor suggestion. What do you think about aligning the label to the left like this?

I'll let @MichaelArestad say for sure, but I think center aligned is fine.

MichaelArestad commented:
Honestly, this would be a good bug to report to the Gutenberg project itself for them to resolve longer block names in the UI.

dereksmart commented:
Marking as in progress -- I plan on splitting this up into two PRs -- one for the build and one for the block

georgestephanis commented on modules/subscriptions.php:
```diff
@@ -1064,6 +1108,13 @@ function jetpack_do_subscription_form( $instance ) {
 	$instance['show_subscribers_total'] = empty( $instance['show_subscribers_total'] ) ? false : true;
 
+	if (
+		array_key_exists( 'subscribe_text', $instance ) &&
```
Would this be more readable as an `isset()` to match the following line?

georgestephanis commented on modules/subscriptions.php:
```diff
@@ -1064,6 +1108,13 @@ function jetpack_do_subscription_form( $instance ) {
+		array_key_exists( 'subscribe_text', $instance ) &&
+		is_array( $instance['subscribe_text'] )
+	) {
+		$instance['subscribe_text'] = $instance['subscribe_text'][0];
```
If it's not numeric array, this could be an issue?  Should this just be an `array_shift()` instead?

oskosk commented:
I've just rebased this one so it can trigger a rebuild in Jetpack Beta for testing

oskosk commented:
I rebased again because JS linting/tests were not running properly.

MichaelArestad commented:
Let's not merge any gutenberg blocks just yet.

oskosk commented:
Closing as this is stalled
